src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -69,47 +73,115 @@ IAsyncServiceProvider serviceProvider
+        {
+            const string eventName = nameof(IVsSolutionRestoreService2) + "." + nameof(NominateProjectAsync);
+            var eventData = new NominateProjectAsyncEventData()
             {
-                Name = projectUniqueName
+                ProjectUniqueName = projectUniqueName
             };
-            dgSpec.AddProject(packageSpec);
-            dgSpec.AddRestore(packageSpec.Name);
-            _projectSystemCache.AddProjectRestoreInfo(projectNames, dgSpec, new List<IAssetsLogMessage>());
+            NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.StartEventOptions, eventData);
+
+            try
+            {
+                Assumes.NotNullOrEmpty(projectUniqueName);
+
+                ProjectNames projectNames = await GetProjectNamesAsync(projectUniqueName, token);
+                var dgSpec = new DependencyGraphSpec();
+                var packageSpec = new PackageSpec()
+                {
+                    Name = projectUniqueName
+                };
+                dgSpec.AddProject(packageSpec);
+                dgSpec.AddRestore(packageSpec.Name);
+                _projectSystemCache.AddProjectRestoreInfo(projectNames, dgSpec, new List<IAssetsLogMessage>());
 
-            await PopulateRestoreInfoSourcesAsync();
-            // returned task completes when scheduled restore operation completes.
-            var restoreTask = _restoreWorker.ScheduleRestoreAsync(
-                SolutionRestoreRequest.OnUpdate(),
-                token);
+                await PopulateRestoreInfoSourcesAsync();
+                // returned task completes when scheduled restore operation completes.
+                var restoreTask = _restoreWorker.ScheduleRestoreAsync(
+                    SolutionRestoreRequest.OnUpdate(),
+                    token);
 
-            return await restoreTask;
+                return await restoreTask;
+            }
+            finally
+            {
+                NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.StopEventOptions);
+            }
+        }
+
+        [System.Diagnostics.Tracing.EventData]
+        private struct NominateProjectAsyncEventData
+        {
+            [System.Diagnostics.Tracing.EventField]
+            public string ProjectUniqueName { get; set; }
         }
 
         public Task<bool> NominateProjectAsync(string projectUniqueName, IVsProjectRestoreInfo projectRestoreInfo, CancellationToken token)
         {
-            return NominateProjectAsync(projectUniqueName, projectRestoreInfo, null, token);
+            const string eventName = nameof(IVsSolutionRestoreService) + "." + nameof(NominateProjectAsync);
+            var eventData = new NominateProjectAsyncEventData()
+            {
+                ProjectUniqueName = projectUniqueName
+            };
+            NuGetExtensibilityEtw.EventSource.Write(eventName, NuGetExtensibilityEtw.StartEventOptions, eventData);
```

Comment:
If this returns a disposable instance, you could reduce the boilerplate together with the new using statement.

Reply:
Granted you'd be allocating a bit there, but may it's not *that* bad.

Reply:
We could use a disposable struct. That would avoid allocations. However, it gets trickier once we consider some events pass event data. I think the event data is extremely useful for problem investigation, so I think we'll lose something valuable if we don't keep that functionality. If we only ever use event data on start events, we can "easily" do it with a second disposable struct that uses generics.  However, I don't have a good idea if we want to enable event data on stop events.

Something like:

```cs
public struct NuGetEtwEvent : IDisposable
{
  private string _eventName;

  public NuGetEtw(string eventName)
  {
    NuGetEtw.Write(eventName, NuGetEtw.StartOptions);
   _eventName = eventName;
  }

  public void Dispose()
  {
    NuGetEtw.Write(eventName, NuGetEtw.StopOptions);
  }
}

public struct NuGetEtwEvent<T> : IDisposable
{
  private string _eventName;

  public NuGetEtw<T>(string eventName, T eventData)
  {
    NuGetEtw.Write(eventName, NuGetEtw.StartOptions, eventData);
   _eventName = eventName;
  }

  public void Dispose()
  {
    NuGetEtw.Write(eventName, NuGetEtw.StopOptions);
  }
}

[Export(typeof(IVsSolutionRestoreServices))]
public VsSolutionRestoreServices : IVsSolutionRestoreServices
{
  [EventDataClass]
  private struct NominateProjectEventData
  {
    [EventDataField]
    public string Project { get; set; }
  }

  public async Task NominateProjectAsync(string projectFile, IVsProjectInfo projectInfo)
  {
    var eventData = new NominateProjectEventData() { Project = projectFile };
    const string eventName = "...";
    using etw = new NuGetEtwEvent(eventName, eventData);

    // rest of code
  }
}
```

Reply:
Do we have a use case for data at the stop event yet?

If no, then I wouldn't worry too much. 

After all it's something that could be solved later on right?

Reply:
Discussed offline. I do have ideas on how it might be useful for add stop event data, but I lack the practical experience with ETW to estimate whether or not we'll actually use it.

An idea to deal with it is to use a `Func<T2>` in the disposable constructor, then call it func in the dispose method and pass the result to `EventSource.Write`. Then we rely on c# compiler anonymous methods automatically capturing scope to be able to use data generating during execution of the instrumented method and return it in the anonymous method. It's spaghetti code when trying to describe it, but the implementation is easy enough to understand:

```cs
public void InstallPackage(DTE.Project project, string packageId, string packageVersion)
{
  const string eventName = "...";
  string projectName = null;
  using etw = new NuGetEtwEvent(eventName:
    startEventData: new
      {
        PackageId = packageId,
        PackageVersion = packageVersion
      },
    endEventData: () =>
      {
        return new {
          Project = projectName
        }
      });

  JTF.Run(async ()=>
    {
      await JTF.SwitchToMainThreadAsync();

      projectName = project.Name;

      // rest of code
    }
}
```

This will no longer be zero allocation, but the majority of our APIs do a lot of allocation for their business logic (install package, uninstall package, get installed packages, etc), so a small number of allocations for the ETW events won't be significant.  We should just be careful for our APIs that are called very often.

Reply:
Update pushed to make all start-stop event pairs use a disposable.

Unfortunately it's not possible to do `using _ = Some.Method()` or `using Some.Method()`, so I had to use `using var _ = Some.Method()` everywhere 😞